src/paddle.cc

```diff
@@ -1178,20 +1268,173 @@ ModelInstanceState::ProcessRequests(
+  } else {
+    memory_type = TRITONSERVER_MEMORY_CPU;
+    memory_type_id = 0;
+  }
+
+  for (uint32_t input_idx = 0; input_idx < input_count; input_idx++) {
+    TRITONBACKEND_Input* input;
+    RETURN_IF_ERROR(
+        TRITONBACKEND_RequestInputByIndex(requests[0], input_idx, &input));
+
+    const char* input_name;
+    TRITONSERVER_DataType input_datatype;
+    const int64_t* input_shape;
+    uint32_t input_dims_count;
+    RETURN_IF_ERROR(TRITONBACKEND_InputProperties(
+        input, &input_name, &input_datatype, &input_shape, &input_dims_count,
+        nullptr, nullptr));
+
+    std::vector<int64_t> batchn_shape;
+    // For a ragged input tensor, the tensor shape should be
+    // the flatten shape of the whole batch
+    if (StateForModel()->IsInputRagged(input_name)) {
+      batchn_shape = std::vector<int64_t>{0};
+      for (size_t idx = 0; idx < request_count; idx++) {
+        TRITONBACKEND_Input* input;
+        RESPOND_AND_SET_NULL_IF_ERROR(
+            &((*responses)[idx]),
+            TRITONBACKEND_RequestInput(requests[idx], input_name, &input));
+        const int64_t* input_shape;
+        uint32_t input_dims_count;
+        RESPOND_AND_SET_NULL_IF_ERROR(
+            &((*responses)[idx]), TRITONBACKEND_InputProperties(
+                                      input, nullptr, nullptr, &input_shape,
+                                      &input_dims_count, nullptr, nullptr));
+
+        batchn_shape[0] += GetElementCount(input_shape, input_dims_count);
+      }
+    }
+    // The shape for the entire input batch, [total_batch_size, ...]
+    else {
+      batchn_shape =
+          std::vector<int64_t>(input_shape, input_shape + input_dims_count);
+      if (max_batch_size != 0) {
+        batchn_shape[0] = total_batch_size;
+      }
+    }
+
+    int64_t batchn_byte_size = GetByteSize(input_datatype, batchn_shape);
+
+    // The input must be in contiguous CPU memory. Use appropriate
+    // allocator info to bind inputs to the right device. .i.e bind inputs
+    // to GPU if they are being provided on GPU.
+    std::string sname(input_name);
+    PD_Tensor* & pd_tensor = input_tensors_[sname];
+    auto pd_shapes = TRITONPADDLE_Shape(batchn_shape).CompatibleShape();
+    PD_TensorReshape(pd_tensor, pd_shapes.size(), pd_shapes.data());
+    char* input_ptr;
+    RETURN_IF_ERROR(PaddleInputMutableData(
+                      pd_tensor, input_datatype,
+                      pd_place_type_, &input_ptr));
+
+    collector->ProcessTensor(input_name, input_ptr, batchn_byte_size,
+                             memory_type, memory_type_id);
+  }
+
+  // Finalize...
+  *cuda_copy |= collector->Finalize();
+  return nullptr;
 }
 
+TRITONSERVER_Error*
+ModelInstanceState::Run(
+    std::vector<TRITONBACKEND_Response*>* responses,
+    const uint32_t response_count)
+{
+  PD_PredictorRun(pd_predictor_);
+#ifdef TRITON_ENABLE_GPU
+  if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+    cudaStreamSynchronize(CudaStream());
```

Comment:
Looks like PD_PredictorRun can be asynchronous for GPU.
Is this synchronization necessary?  Can't you just proceed with the GPU copyout and let synchronization happen in ReadOutputTensors ? Because the copyout will happen on the same stream the copies will be ordered.
You can use cuda events to get accurate compute_infer and compute_output times.
Minimizing extra synchronization might have significant performance impact on very low latency models.

Reply:
Yes, PD_PredictorRun can be asynchronous for GPU.